constantine-rust/constantine-ethereum-evm-precompiles/tests/t_ethereum_evm_precompiles.rs

```diff
@@ -89,13 +89,49 @@ fn from_hex(hex: HexString) -> Result<Vec<u8>, bool> {
     // TODO: why cannot use `with_capacity`?
     let mut result = vec![0u8; href.len() / 2];
     match hex::decode_to_slice(href, result.as_mut_slice() as &mut [u8]) {
-        Ok(_) => Ok(result),
-        Err(_) => Err(false),
+        Ok(_) => Some(result),
+        Err(_) => None,
+    }
+}
+
+trait CompareReturn {
+    fn compare(self, v: Vec<u8>) -> bool;
+}
+
+impl CompareReturn for [u8; 32] {
+    fn compare(self, v: Vec<u8>) -> bool {
+        return self.to_vec() == v;
+    }
+}
+
+impl CompareReturn for [u8; 64] {
+    fn compare(self, v: Vec<u8>) -> bool {
+        return self.to_vec() == v;
+    }
+}
+
+impl CompareReturn for [u8; 128] {
+    fn compare(self, v: Vec<u8>) -> bool {
+        return self.to_vec() == v;
     }
 }
 
+impl CompareReturn for [u8; 256] {
+    fn compare(self, v: Vec<u8>) -> bool {
+        return self.to_vec() == v;
+    }
+}
 
-fn t_generate(test_name: String, func: TestFunction) {
+impl CompareReturn for Vec<u8> {
+    fn compare(self, v: Vec<u8>) -> bool {
+        return self == v;
+    }
+}
+
+fn t_generate<T: CompareReturn, F>(test_name: String, func: F)
```

Comment:
```suggestion
fn t_generate<T: impl Iterator<Item = u8>, F>(test_name: String, func: F)
```

here

Reply:
While your proposal is not quite correct (`impl Iterator` is not valid syntax must be without the `impl`), it led me down the right path. `Iterator` would require the test functions to actually return an iterator. But there's also a predefined `IntoIterator` trait, which works (via `x.into_iter().collect::<Vec<u8>>()`).

Reply:
Ah yes,  the impl Iterator is for arguments, not generics. And yes Vec use `into_iter` indeed.